### PR TITLE
provider: online

### PR DESCRIPTION
**Pull request by kapouer**
Hi, this a `online` provider (a.k.a. online.net, scaleway, dedibox...).

Feel free to comment while i still care !

### CODE REVIEW

**Comment by CapsuleCD** [comment]
Hi.

I'm an automated pull request bot named [CapsuleCD](https://www.capsulecd.com). I handle testing, versioning and package releases for this project. 
- If you're the owner of this repo, you can click the button below to kick off the CapsuleCD build pipeline and create an automated release.'
- If not, don't worry, someone will be by shortly to check on your pull request. 

[CapsuleCD](https://www.capsulecd.com/project/github/AnalogJ/lexicon/pullrequests/239)

---
If you're interested in learning more about [CapsuleCD](http://www.github.com/AnalogJ/capsulecd), or adding it to your project, you can check it out [here](https://www.capsulecd.com)

**Comment by kapouer** [comment]
Here it is !
(Squashed in a single commit)
I did not understand (not being a python dev) how to run the tests (only with that provider), but i manually tested obvious things like create/update/delete A or TXT.

**Comment by AnalogJ** [comment]
Hey @kapouer 
Thanks for the PR! 

Looks like the only failure is in the transip provider, which @adferrand  just submitted a fix for. 
Lets get your PR retested and verify that its working correctly.

**Comment by AnalogJ** [comment]
I dont see any test recordings, so you'll need to commit those if possible. 
The instructions for how to make recordings are included in the `Contributing.md` file.

**Comment by CapsuleCD** [comment]
Hi.

I'm an automated pull request bot named [CapsuleCD](https://www.capsulecd.com). I handle testing, versioning and package releases for this project. 
- If you're the owner of this repo, you can click the button below to kick off the CapsuleCD build pipeline and create an automated release.'
- If not, don't worry, someone will be by shortly to check on your pull request. 

[CapsuleCD](https://www.capsulecd.com/project/github/AnalogJ/lexicon/pullrequests/239)

---
If you're interested in learning more about [CapsuleCD](http://www.github.com/AnalogJ/capsulecd), or adding it to your project, you can check it out [here](https://www.capsulecd.com)

**Comment by AnalogJ** [comment]
I also noticed that you're missing a test configuration file, and API recordings so that the `Online` provider will be automatically tested by the test suite.

Without those, I will be unable to merge this PR. 

The instructions are available in the CONTRIBUTING.md file. These two sections will probably be the most valuable:
- [Testing Your Provider](https://github.com/AnalogJ/lexicon/blob/master/CONTRIBUTING.md#testing-your-provider)
- [Test Recordings](https://github.com/AnalogJ/lexicon/blob/master/CONTRIBUTING.md#test-recordings)

If you need any help, I can walk you though any issues you may have.

**Comment by kapouer** [comment]
So if i understand it right, the api token needed to run the tests will be kept known to you and you only ?

**Comment by AnalogJ** [comment]
Hey @kapouer 
Actually no. The API token can be yours, and then the HTTP request and response is recorded by `vcrpy` and stored in https://github.com/AnalogJ/lexicon/tree/master/tests/fixtures/cassettes

When the tests are run by CircleCI or any user, those recordings are used, instead of talking to the API directly, meaning no API token is necessary.

You'll need to ensure that any authentication headers for that provider are removed from the recordings before committing. Heres and example of how I did that with Cloudflare:

https://github.com/AnalogJ/lexicon/blob/master/tests/providers/test_cloudflare.py#L15-L16

Does that all make sense?

**Comment by kapouer** [comment]
Not at all ! I throw the towel now and think `vcrpy` is waaay above the top of my head.

**Comment by kapouer** [comment]
I'll just mimic the other providers without asking questions :)

**Comment by kapouer** [comment]
The `create_record_for_A_with_valid_name_and_content` test (and all tests, actually, but this is the first one) pass `placeholder_priority` as priority (here's what the provider receives: `{'priority': 'placeholder_priority', 'data': '127.0.0.1', 'type': 'A', 'name': 'localhost.edasarl.fr.', 'ttl': 3600}`).
I don't think this is right... It should pass nothing or an integer.

**Comment by adferrand** [comment]
Parameter priority seems to be specific to your provider. Then you should override method _test_engine_overrides in your test suite class, as described in the method documentation from integration_tests, to provide a good default value for priority.

**Comment by AnalogJ** [comment]
By default the `fallbackFn` returns `None`.
However, during testing it will return a `placeholder_` prefixed value to mimic data passed via environmental variables. 

If you want to set a value/default for `priority` then just add the following block in your test file

```
def _test_options(self):
    options = super(OnlineProviderTests, self)._test_options()
    options.update({'provider':'10'}) 
    return options

```

**Comment by AnalogJ** [comment]
If you commit your provider test file I can take a look and ensure it works correctly.

**Comment by adferrand** [comment]
@AnalogJ what is the difference between _test_engine_overrides and _test_options ? Both methods seems to have the same result, provide default value for parameters during tests.

**Comment by AnalogJ** [comment]
Basically its the result of a partly complicated options system. It definitely needs a bit more documentation. 

The relevant runtime source is [here](https://github.com/AnalogJ/lexicon/blob/703b55f27619e68dcc1e207523d977b6f78aff09/lexicon/providers/base.py#L34-L40), test time source is [here](https://github.com/AnalogJ/lexicon/blob/master/tests/providers/integration_tests.py#L316-L328).

Basically, there are 4 types of configuration options
- static defaults - set on every provider, in test and runtime. right now theres only one, `{'ttl': 3600}`.
- engine overrides - used for overriding engine specific options during test time (eg. api endpoint). Empty at runtime. 
- env options - options specified by environmental variables (must be prefixed by `LEXICON_AUTH`). meant for credentials/tokens
- cli options - any options/flags specified via command line parameters (empty at during testing)

**Comment by kapouer** [comment]
it's going better and better... i'm fixing new failures now. Thanks for the help !

**Comment by kapouer** [comment]
I had to work around `test_Provider_when_calling_update_record_should_modify_record` which relies on finding a record by id to return True even if the record was not updated (because it could not find the record by id).
Otherwise tests pass now.

File: tests/providers/test_online.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	def _filter_headers(self):
+		return ['Authorization']
+
+	def _test_engine_overrides(self):
+		overrides = super(OnlineProviderTests, self)._test_engine_overrides()
+		overrides['fallbackFn'] = (lambda x: 'placeholder_' + x if x != 'priority' else '')
+		return overrides
+
+	@pytest.fixture(autouse=True)
+	def skip_suite(self, request):
```
Comment (AnalogJ) [inline tests/providers/test_online.py]: Hey, This PR looks just about perfect but it seems you're skipping the extended suite 
Can you remove lines 24 - 27 and ensure that your provider supports record sets?

Comment (kapouer) [reply tests/providers/test_online.py]: Not that i don't want to, but i'm confused about the meaning of record sets.
In this provider, one can only modify a *disabled* record set, and need to activate it to make it live.
This means, in particular, that the id of a given record will change when the record set it is being part of is made active.
Is this compatible with how lexicon deal with things ?

Comment (kapouer) [reply tests/providers/test_online.py]: @lexicon maybe i replied that fast that you did not notice :)

Comment (AnalogJ) [reply tests/providers/test_online.py]: Looking at the documentation, it looks like only zones can be enabled/disabled, not record sets. 

fundamentally the request is that you can have multiple records for same name & type, with differing content.




Type | Name | Content
------------ | ------------- | -------------
A | test1.example.com | 1.2.3.4 
A | test2.example.com | 4.3.2.1
TXT | _acme-challenge.example.com | D-52Wm4V7xoUpGax-F8FrPO45cQRcbRj-XoblaY4uYM
TXT | _acme-challenge.example.com | 6lOgCI0p_LRhtrJMh9aTYAek6hZ64nT75-DkeeQccfA

Comment (kapouer) [reply tests/providers/test_online.py]: oh right, indeed if record set is only that feature, it's easy to support.

**Comment by AnalogJ** [comment]
Hey @kapouer 
How are the changes to support record sets going? Anything I can help with?

**Comment by kapouer** [comment]
Was busy this week ! I'll give a shot in an hour or two.

**Comment by kapouer** [comment]
Extended tests enabled, rebased, sensitive data removed, squashed in one commit.

**Comment by AnalogJ** [comment]
awesome work. I'll get this merged and released right away.

**Pull request merged**